README.md
# Paylike Dart API client

_This is an ALPHA release_

Although the functionality included is production quality, the supported scope
of the API is merely a stub.

High-level client for the API documented at:
https://github.com/paylike/api-reference. It is using
[paylike_dart_request](https://pub.dev/packages/paylike_dart_request) under the
hood.

## Examples

```sh
dart pub add paylike_dart_client
```

```dart
import 'package:paylike_dart_client/paylike_dart_client.dart';

void main() {
  var client = PaylikeClient('MY_CLIENT_ID');

  PaylikeRequestBuilder<TokenizedResponse> request = client.tokenize(TokenizeTypes.PCN, '1000000000000000');
  request.execute().then((response) {
    print('Received token: ' + response.token);
  }).catchError((e) => print(e));
}

```

## Methods

```dart
// client.tokenize(TokenizeTypes.PCN, '...');
// client.tokenize(TokenizeTypes.PCSC, '...');
PaylikeRequestBuilder<TokenizedResponse> tokenize(TokenizeTypes type, String value);

// client.paymentCreate(payment);
PaylikeRequestBuilder<PaymentResponse> paymenCreate({
  required Map<String, dynamic> payment,
  List<String> hints = const [],
});
```

[More information](https://github.com/paylike/api-reference/blob/main/payments/index.md) on payment data structure.

## Error handling

The methods may throw any error forwarded from the used PaylikeRequester implementation as
well as one of the below error classes.

- `RateLimitException`

  May have a `retryAfter` (Duration) property if sent by the server
  specifying the minimum delay.

- `TimeoutException`

  Comes from `dart:async` library https://api.dart.dev/be/169657/dart-async/TimeoutException-class.html

- `ServerErrorException`

  Has `status` and `headers` properties copied from the io.HttpClientResponse

- `PaylikeException`

  These errors correspond to
  [status codes](https://github.com/paylike/api-reference/blob/master/status-codes.md)
  from the API reference. They have at least a `code` and `message` property,
  but may also have other useful properties relevant to the specific error code,
  such as a minimum and maximum for amounts.

## Logging

Pass a log function of the format `void Function(dynamic d)` to catch internal (structured)
logging.

```dart
  var client = PaylikeClient('MY_CLIENT_ID').setLog((dynamic d) => print(d))
```

## Timeouts and retries

There is a default timeout for all HTTPS requests of 10 seconds and a retry
strategy of 10 retries with increasing delay. 
```dart
      switch (attempts) {
        case 0:
        case 1:
          usedDuration = Duration(milliseconds: 0);
          break;
        case 2:
          usedDuration = Duration(milliseconds: 100);
          break;
        case 3:
          usedDuration = Duration(seconds: 2);
          break;
        default:
          usedDuration = Duration(seconds: 10);
      }
```

Using the default retry handler is recommended which you can do by specifying on the `PaylikeRequestBuilder` you receive:
```dart
      var request = client
          .tokenize(TokenizeTypes.PCN, '4100000000000000')
          .withDefaultRetry();
      var response = await request.execute();
```
You can also create your own handler by implementing the RetryHandler abstract class:
```dart
class CustomRetryHandler<T> implements RetryHandler<T> {
  int attempts = 0;
  @override
  Future<T> retry(Future<T> Function() executor) async {
    try {
      var res = await executor();
      return res;
    } on RateLimitException catch (e) {
      attempts++;
      if (attempts > 10) {
        rethrow;
      }
      await Future.delayed(Duration(seconds: 5));
    } catch (e) {
      rethrow;
    }
    return retry(executor);
  }
}

// Then you can apply your own retry handler to the request builder:
var request = client
      .tokenize(TokenizeTypes.PCN, '4100000000000000')
      .withRetry(CustomRetryHandler<TokenizedResponse>());
var response = await request.execute();
```


CHANGELOG.md
## 1.0.10

- Refactored codebase to allow exports to work modularly

## 1.0.9

- Fixed a typo in the previous release

## 1.0.8

- Added option to access Apple Pay

## 1.0.7

- Fixed an issue with PaylikeException not being exported from `paylike_dart_request`

## 1.0.6

- Fixed issues with interface documentation

## 1.0.5

- Fixed an issue with default constructor not allowing empty client ID

## 1.0.4

- Fixed an issue with x-client header not being complient to the rest of the Paylike ecosystem

## 1.0.3

- Added response logging
- Fixed an issue with hints not being properly constructed for TDS
- Updated paylike_dart_request dependency
- Added a simple fallback if the request encounters a natural timeout, now it should at least try once more

## 1.0.2

- Changed paymentCreate interface for better usability

## 1.0.1

- Added support for iframe challenges

## 1.0.0

- Initial implementation


test/paylike_dart_client_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:mockito/annotations.dart';
import 'package:mockito/mockito.dart';
import 'package:paylike_dart_client/paylike_dart_client.dart';
import 'package:paylike_dart_client/src/dto.dart';
import 'package:paylike_dart_client/src/request.dart';
import 'package:paylike_dart_request/paylike_dart_request.dart';
import 'package:test/test.dart';

import 'paylike_dart_client_test.mocks.dart';

class Mocker {
  var response = MockPaylikeResponse();
  var requester = MockPaylikeRequester();
}

var E2E_TESTING_ENABLED = Platform.environment['E2E_TEST'] == 'true';
var E2E_CLIENT_KEY = Platform.environment['E2E_CLIENT_KEY'];

@GenerateMocks([PaylikeRequester, PaylikeResponse])
void main() {
  group('Essential tests', () {
    test('Tokenization should be able to provide back a token', () async {
      var mocker = Mocker();
      var client = PaylikeClient().setRequester(mocker.requester);
      when(mocker.requester.request(any, any))
          .thenAnswer((realInvocation) async => mocker.response);
      when(mocker.response.getBody())
          .thenAnswer((realInvocation) async => '{"token":"foo"}');
      var request = client
          .tokenize(TokenizeTypes.PCN, '4100000000000000')
          .withDefaultRetry();
      var resp = await request.execute();
      expect(resp.token, 'foo');
    });

    test('Payment creation should do challenges based on fetch type', () async {
      var mocker = Mocker();
      var counter = 0;
      var client = PaylikeClient().setRequester(mocker.requester);
      var challengeResolvedResponse = MockPaylikeResponse();
      var finalResolveResponse = MockPaylikeResponse();
      when(finalResolveResponse.getBody())
          .thenAnswer((realinvocation) async => '{"authorizationId": "foo"}');
      when(mocker.requester.request(client.hosts.api + '/payments', any))
          .thenAnswer((realInvocation) async {
        if (counter == 0) {
          counter++;
          return mocker.response;
        }
        var opts = realInvocation.positionalArguments[1] as RequestOptions;
        expect((opts.data as Map<String, dynamic>)['test'], {});
        expect((opts.data as Map<String, dynamic>)['hints'], ['hint1']);
        return finalResolveResponse;
      });
      when(mocker.response.getBody())
          .thenAnswer((realInvocation) async => jsonEncode({
                'challenges': [
                  {'name': 'challenge01', 'type': 'fetch', 'path': '/challenge'}
                ]
              }));
      when(mocker.requester.request(client.hosts.api + '/challenge', any))
          .thenAnswer((realInvocation) async => challengeResolvedResponse);
      when(challengeResolvedResponse.getBody())
          .thenAnswer((realInvocation) async => jsonEncode({
                'hints': ['hint1']
              }));
      var resp = await client
          .paymentCreate(payment: {'test': {}})
          .withDefaultRetry()
          .execute();
      expect(resp.paymentResponse, isNotNull);
      expect(resp.isHTML, false);
      expect(resp.paymentResponse?.transaction.id, 'foo');
    });

    test('Default retry mechanism should work as expected with tokenize',
        () async {
      var retryHandler = DefaultRetryHandler<TokenizedResponse>();
      try {
        await retryHandler.retry(() async {
          throw RateLimitException.withTime('200');
        });
        fail('Should not be able to reach this');
      } catch (e) {
        expect(retryHandler.attempts, 11);
        expect(e is RateLimitException, true);
      }
      retryHandler = DefaultRetryHandler<TokenizedResponse>();
      try {
        var count = 0;
        var resp = await retryHandler.retry(() async {
          count++;
          if (count == 11) {
            return TokenizedResponse.fromJSON({'token': 'foo'});
          }
          throw RateLimitException.withTime('200');
        });
        expect(retryHandler.attempts, 10);
        expect(resp.token, 'foo');
      } catch (e) {
        fail('Should not be able to reach this');
      }
    });
  });

  group('End to end tests', () {
    if (E2E_CLIENT_KEY == null || E2E_CLIENT_KEY!.isEmpty) {
      throw Exception('E2E_CLIENT_KEY is required for E2E tests');
    }
    final client = PaylikeClient();

    test('Tokenization should work as expected', () async {
      var request = client
          .tokenize(TokenizeTypes.PCN, '4100000000000000')
          .withDefaultRetry();
      var response = await request.execute();
      expect(response.token, isNotNull);
    });

    test('Payment creation should work as expected', () async {
      String cardNumber;
      String cardCode;
      {
        var request = client
            .tokenize(TokenizeTypes.PCN, '4100000000000000')
            .withDefaultRetry();
        var response = await request.execute();
        cardNumber = response.token;
      }
      {
        var request =
            client.tokenize(TokenizeTypes.PCSC, '111').withDefaultRetry();
        var response = await request.execute();
        cardCode = response.token;
      }
      var response = await client
          .paymentCreate(payment: {
            'test': {},
            'integration': {
              'key': E2E_CLIENT_KEY as String,
            },
            'amount': {
              'currency': 'EUR',
              'value': 1000,
              'exponent': 0,
            },
            'card': {
              'number': {
                'token': cardNumber,
              },
              'code': {
                'token': cardCode,
              },
              'expiry': {'month': 12, 'year': 2022},
            },
          })
          .withDefaultRetry()
          .execute();
      expect(response.isHTML, true);
      expect(response.HTMLBody, isNotNull);
      expect(response.hints, isNotEmpty);
      expect(response.paymentResponse, isNull);
    });
  }, skip: !E2E_TESTING_ENABLED);
}


test/paylike_dart_client_test.mocks.dart
// Mocks generated by Mockito 5.0.15 from annotations
// in paylike_dart_client/test/paylike_dart_client_test.dart.
// Do not manually edit this file.

import 'dart:async' as _i4;
import 'dart:io' as _i2;

import 'package:mockito/mockito.dart' as _i1;
import 'package:paylike_dart_request/src/paylike_dart_request_base.dart' as _i3;

// ignore_for_file: avoid_redundant_argument_values
// ignore_for_file: avoid_setters_without_getters
// ignore_for_file: comment_references
// ignore_for_file: implementation_imports
// ignore_for_file: invalid_use_of_visible_for_testing_member
// ignore_for_file: prefer_const_constructors
// ignore_for_file: unnecessary_parenthesis

class _FakeHttpClient_0 extends _i1.Fake implements _i2.HttpClient {}

class _FakePaylikeRequester_1 extends _i1.Fake implements _i3.PaylikeRequester {
}

class _FakePaylikeResponse_2 extends _i1.Fake implements _i3.PaylikeResponse {}

class _FakeHttpClientResponse_3 extends _i1.Fake
    implements _i2.HttpClientResponse {}

/// A class which mocks [PaylikeRequester].
///
/// See the documentation for Mockito's code generation for more information.
class MockPaylikeRequester extends _i1.Mock implements _i3.PaylikeRequester {
  MockPaylikeRequester() {
    _i1.throwOnMissingStub(this);
  }

  @override
  Function get log =>
      (super.noSuchMethod(Invocation.getter(#log), returnValue: () {})
          as Function);
  @override
  set log(Function? _log) => super.noSuchMethod(Invocation.setter(#log, _log),
      returnValueForMissingStub: null);
  @override
  _i2.HttpClient get client => (super.noSuchMethod(Invocation.getter(#client),
      returnValue: _FakeHttpClient_0()) as _i2.HttpClient);
  @override
  set client(_i2.HttpClient? _client) =>
      super.noSuchMethod(Invocation.setter(#client, _client),
          returnValueForMissingStub: null);
  @override
  _i3.PaylikeRequester setLog(Function? log) =>
      (super.noSuchMethod(Invocation.method(#setLog, [log]),
          returnValue: _FakePaylikeRequester_1()) as _i3.PaylikeRequester);
  @override
  _i3.PaylikeRequester setClient(_i2.HttpClient? client) =>
      (super.noSuchMethod(Invocation.method(#setClient, [client]),
          returnValue: _FakePaylikeRequester_1()) as _i3.PaylikeRequester);
  @override
  _i4.Future<_i3.PaylikeResponse> request(
          String? endpoint, _i3.RequestOptions? opts) =>
      (super.noSuchMethod(Invocation.method(#request, [endpoint, opts]),
              returnValue:
                  Future<_i3.PaylikeResponse>.value(_FakePaylikeResponse_2()))
          as _i4.Future<_i3.PaylikeResponse>);
  @override
  String toString() => super.toString();
}

/// A class which mocks [PaylikeResponse].
///
/// See the documentation for Mockito's code generation for more information.
class MockPaylikeResponse extends _i1.Mock implements _i3.PaylikeResponse {
  MockPaylikeResponse() {
    _i1.throwOnMissingStub(this);
  }

  @override
  _i2.HttpClientResponse get response =>
      (super.noSuchMethod(Invocation.getter(#response),
          returnValue: _FakeHttpClientResponse_3()) as _i2.HttpClientResponse);
  @override
  set response(_i2.HttpClientResponse? _response) =>
      super.noSuchMethod(Invocation.setter(#response, _response),
          returnValueForMissingStub: null);
  @override
  _i4.Future<String> getBody() =>
      (super.noSuchMethod(Invocation.method(#getBody, []),
          returnValue: Future<String>.value('')) as _i4.Future<String>);
  @override
  _i4.Future<_i4.Stream<dynamic>> getBodyReader() =>
      (super.noSuchMethod(Invocation.method(#getBodyReader, []),
              returnValue:
                  Future<_i4.Stream<dynamic>>.value(Stream<dynamic>.empty()))
          as _i4.Future<_i4.Stream<dynamic>>);
  @override
  String toString() => super.toString();
}


lib/src/dto.dart
/// PaymentChallenge describes a challenge after a payment creation
/// is initiated
class PaymentChallenge {
  late String name;
  late String type;
  late String path;
  PaymentChallenge.fromJSON(Map<String, dynamic> json)
      : name = json['name'],
        type = json['type'],
        path = json['path'];
}

/// Describes the hints array received after executing a challenge successfully.
class HintsResponse {
  late List<String> hints;
  HintsResponse.fromJSON(Map<String, dynamic> json)
      : hints = (json['hints'] as List<dynamic>).cast<String>();
}

/// Describes a response from tokenize.
class TokenizedResponse {
  String token;
  TokenizedResponse.fromJSON(Map<String, dynamic> json) : token = json['token'];
}

/// Describes a paylike transaction.
class PaylikeTransaction {
  String id;
  PaylikeTransaction(this.id);
  PaylikeTransaction.fromJSON(Map<String, dynamic> json)
      : id = json['authorizationId'] ?? json['transactionId'];
}

/// Describes a payment response.
class PaymentResponse {
  PaylikeTransaction transaction;
  Map<String, dynamic>? custom;
  PaymentResponse.fromJSON(Map<String, dynamic> json)
      : transaction = PaylikeTransaction.fromJSON(json),
        custom = json['custom'];
}

/// Describes the client response from the Paylike capture API
class PaylikeClientResponse {
  final bool isHTML;
  PaymentResponse? paymentResponse;
  List<String> hints;
  String? HTMLBody;
  PaylikeClientResponse({
    required this.isHTML,
    this.paymentResponse,
    this.HTMLBody,
    this.hints = const [],
  });

  /// Returns the payment response if not null
  /// otherwise throws an exception
  PaymentResponse getPaymentResponse() {
    if (paymentResponse == null) {
      throw Exception('Payment response is null, cannot be acquired');
    }
    return paymentResponse as PaymentResponse;
  }

  /// Returns HTML body if not null
  /// otherwise throws an exception
  String getHTMLBody() {
    if (HTMLBody == null) {
      throw Exception('HTMLBody is null, cannot be acquired');
    }
    return HTMLBody as String;
  }
}


lib/src/request.dart
import 'dart:async';

import 'package:paylike_dart_request/paylike_dart_request.dart';

/// RetryHandler describes the interfaces of a retry handler.
abstract class RetryHandler<T> {
  /// Retry is the function that should be implemented by every retry handler.
  Future<T> retry(Future<T> Function() executor);
}

/// DefaultRetryHandler is used as the default retry backoff
/// mechanism for handling RateLimitExceptions.
class DefaultRetryHandler<T> implements RetryHandler<T> {
  /// Counts the number of attempts made so far
  int attempts = 0;

  /// Gives back the duration suggested by the API
  /// or a Duration based on the number of attempts if no
  /// retry headers were provided.
  Duration getRetryAfter(Duration? retryAfter) {
    var usedDuration = retryAfter ?? Duration(milliseconds: 0);
    if (retryAfter == null) {
      switch (attempts) {
        case 0:
        case 1:
          usedDuration = Duration(milliseconds: 0);
          break;
        case 2:
          usedDuration = Duration(milliseconds: 100);
          break;
        case 3:
          usedDuration = Duration(seconds: 2);
          break;
        default:
          usedDuration = Duration(seconds: 10);
      }
    }
    return usedDuration;
  }

  /// Implementation of the retry mechanism.
  @override
  Future<T> retry(Future<T> Function() executor) async {
    try {
      var res = await executor();
      return res;
    } on RateLimitException catch (e) {
      attempts++;
      if (attempts > 10) {
        rethrow;
      }
      await Future.delayed(getRetryAfter(e.retryAfter));
      return retry(executor);
    } on TimeoutException catch (_) {
      attempts++;
      if (attempts > 2) {
        rethrow;
      }
      return retry(executor);
    } catch (e) {
      rethrow;
    }
  }
}

/// PaylikeRequestBuilder provides a flexible way to add
/// custom retry mechanism into the flow.
class PaylikeRequestBuilder<T> {
  /// Async function to execute
  late Future<T> Function() fn;

  /// Handler to do retry
  RetryHandler<T> retryHandler = DefaultRetryHandler<T>();

  /// Indicates if retry is enabled
  bool retryEnabled = false;
  PaylikeRequestBuilder(this.fn);

  /// Receives a custom retry implementation and enables retry mechanism.
  PaylikeRequestBuilder<T> withRetry(RetryHandler<T> retryHandler) {
    this.retryHandler = retryHandler;
    retryEnabled = true;
    return this;
  }

  /// Enables default retry - backoff mechanism.
  PaylikeRequestBuilder<T> withDefaultRetry() {
    retryEnabled = true;
    return this;
  }

  /// Executes the request.
  Future<T> execute() {
    return retryEnabled ? retryHandler.retry(fn) : fn();
  }
}


lib/src/paylike_dart_client_base.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';
import 'package:paylike_dart_request/paylike_dart_request.dart';

import 'dto.dart';
import 'request.dart';

/// Describes endpoints used.
class PaylikeHosts {
  String api = 'https://b.paylike.io';
  String vault = 'https://vault.paylike.io';
  String applePayAPI = 'https://applepay.paylike.io/token';
  PaylikeHosts();

  /// Creates a new PaylikeHosts from two given URLs.
  PaylikeHosts.from(this.api, this.vault);
}

/// TokenizeTypes describe the options for tokenizing card number and code
enum TokenizeTypes {
  /// For card number
  PCN,

  /// For CVC codes
  PCSC,
}

/// Handles high level requests towards the paylike ecosystem
class PaylikeClient {
  /// Generates a new client ID
  static String _generateClientID() {
    const _chars =
        'AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz1234567890';
    var _rnd = Random();
    String getRandomString(int length) =>
        String.fromCharCodes(Iterable.generate(
            length, (_) => _chars.codeUnitAt(_rnd.nextInt(_chars.length))));
    return 'dart-1-${getRandomString(6)}';
  }

  /// SDK Client ID used while making requests
  ///
  /// Note: This is not the merchant's client ID
  String clientId = _generateClientID();

  /// For testing purposes
  PaylikeClient.withSpecificClientId(this.clientId);
  PaylikeClient();

  /// Logger function
  Function log = (dynamic o) => print('''
  ${jsonEncode(o)}
  ''');

  /// Underlying implementation to do requests
  PaylikeRequester requester = PaylikeRequester();

  /// Timeout to use while making requests
  Duration timeout = Duration(seconds: 20);

  /// Host APIs
  PaylikeHosts hosts = PaylikeHosts();

  /// Overrides the used requester.
  PaylikeClient setRequester(PaylikeRequester requester) {
    this.requester = requester;
    return this;
  }

  /// Overrides the used logger.
  PaylikeClient setLog(void Function(dynamic d) log) {
    this.log = log;
    return this;
  }

  /// Overrides the timeout settings.
  PaylikeClient setTimeout(Duration timeout) {
    this.timeout = timeout;
    return this;
  }

  /// Overrides hosts.
  PaylikeClient setHosts(PaylikeHosts hosts) {
    this.hosts = hosts;
    return this;
  }

  /// Tokenize is used to acquire tokens from the vault
  /// with retry mechanism used.
  PaylikeRequestBuilder<TokenizedResponse> tokenize(
      TokenizeTypes type, String value) {
    return PaylikeRequestBuilder(() => _tokenize(type, value));
  }

  /// Used for the tokenization of a token acquired during apple payment flow
  PaylikeRequestBuilder<TokenizedResponse> tokenizeApple(String value) {
    return PaylikeRequestBuilder(() => _tokenizeApple(value));
  }

  /// Used to acquire tokens from the vault.
  Future<TokenizedResponse> _tokenize(TokenizeTypes type, String value) async {
    var opts = RequestOptions.fromClientId(clientId)
        .setData({
          'type': type == TokenizeTypes.PCN ? 'pcn' : 'pcsc',
          'value': value,
        })
        .setVersion(1)
        .setTimeout(timeout);
    var response = await requester.request(hosts.vault, opts);
    var body = await response.getBody();
    return TokenizedResponse.fromJSON(jsonDecode(body));
  }

  /// Used to acquire tokens for apple pay
  Future<TokenizedResponse> _tokenizeApple(String value) async {
    var opts = RequestOptions.fromClientId(clientId)
        .setData({
          'token': value,
        })
        .setVersion(1)
        .setTimeout(timeout);
    var response = await requester.request(hosts.applePayAPI, opts);
    var body = await response.getBody();
    return TokenizedResponse.fromJSON(jsonDecode(body));
  }

  /// Payment create calls the capture API
  /// with retry mechanism used
  PaylikeRequestBuilder<PaylikeClientResponse> paymentCreate({
    required Map<String, dynamic> payment,
    List<String> hints = const [],
  }) {
    return PaylikeRequestBuilder(() => _paymentCreate(payment, hints, null));
  }

  /// Payment create calls the payment API
  Future<PaylikeClientResponse> _paymentCreate(Map<String, dynamic> payment,
      List<String> hints, String? challengePath) async {
    var subPath = challengePath ?? '/payments';
    var url = hosts.api + subPath;
    var opts = RequestOptions.fromClientId(clientId)
        .setData({
          ...payment,
          'hints': hints,
        })
        .setVersion(1)
        .setTimeout(timeout);
    var response = await requester.request(url, opts);
    Map<String, dynamic> body = jsonDecode(await response.getBody());
    log({
      't': 'response-body',
      'data': body,
    });
    if (body['challenges'] != null &&
        (body['challenges'] as List<dynamic>).isNotEmpty) {
      var remainingChallenges = (body['challenges'] as List<dynamic>)
          .map((e) => PaymentChallenge.fromJSON(e));
      var fetchChallenges = remainingChallenges.where((c) => c.type == 'fetch');
      if (fetchChallenges.isNotEmpty) {
        return _paymentCreate(payment, hints, fetchChallenges.first.path);
      }
      var tdsChallenges = remainingChallenges.where(
          (c) => c.type == 'background-iframe' && c.name == 'tds-fingerprint');
      if (tdsChallenges.isNotEmpty) {
        return _paymentCreate(payment, hints, tdsChallenges.first.path);
      }
      return _paymentCreate(payment, hints, remainingChallenges.first.path);
    }
    // IFRAME
    if (body['action'] != null && body['fields'] != null) {
      var refreshedHints = hints;
      if (body['hints'] != null) {
        refreshedHints =
            <String>{...hints, ...HintsResponse.fromJSON(body).hints}.toList();
      }
      var formResp = await requester.request(
          Uri.parse(body['action']).toString(),
          RequestOptions(
              form: true,
              formFields: (body['fields'] as Map<String, dynamic>)
                  .map((key, value) => MapEntry(key, value.toString()))));
      return PaylikeClientResponse(
          isHTML: true,
          HTMLBody: await formResp.getBody(),
          hints: refreshedHints);
    }
    if (body['hints'] != null && (body['hints'] as List<dynamic>).isNotEmpty) {
      var hintsResp = HintsResponse.fromJSON(body);
      return _paymentCreate(
          payment, <String>{...hints, ...hintsResp.hints}.toList(), null);
    }
    return PaylikeClientResponse(
        isHTML: false, paymentResponse: PaymentResponse.fromJSON(body));
  }
}


example/paylike_dart_client_example.dart
// ignore_for_file: unused_local_variable

import 'package:paylike_dart_client/paylike_dart_client.dart';

void main() {
  var client = PaylikeClient()
      .setHosts(
          PaylikeHosts.from('https://b.paylike.io', 'https://vault.paylike.io'))
      .setLog((dynamic a) => print(a))
      .setTimeout(Duration(seconds: 20));

  () async {
    String cardNumberToken;
    String cardCodeToken;
    {
      var request = client
          .tokenize(TokenizeTypes.PCN, '4100000000000000') // Card Number
          .withDefaultRetry();
      var response = await request.execute();
      cardNumberToken = response.token;
    }
    {
      var request = client
          .tokenize(TokenizeTypes.PCSC, '111') // Card Number
          .withDefaultRetry();
      var response = await request.execute(); // CVC
      cardCodeToken = response.token;
    }
    {
      var payment = {
        'integration': {
          'key': 'MERCHANT_CLIENT_ID',
        },
        'amount': {
          'currency': 'EUR',
          'value': 1000,
          'exponent': 0,
        },
        'card': {
          'number': {
            'token': cardNumberToken,
          },
          'code': {
            'token': cardCodeToken,
          },
          'expiry': {'month': 12, 'year': 2022},
        },
      };
      var request = client.paymentCreate(payment: payment).withDefaultRetry();
      var response = await request.execute();
      // If the response is HTML, then you encountered a payment challenge
      // like TDS
      if (response.isHTML) {
        var htmlBody = response.getHTMLBody();
        var collectedHints = response.hints;
        // ... Handle TDS
        request = client
            .paymentCreate(payment: payment, hints: collectedHints)
            .withDefaultRetry();
        // Contineu after auth
      } else {
        var paymentResp = response.getPaymentResponse();
        print('Acquried transaction reference: ' + paymentResp.transaction.id);
        // ... Store transaction ID
      }
    }
  }()
      .then((_) => print('payment created'))
      .catchError((e) => print(e));
}
